added config argument for oauth2.0 and init_oauth

muschellij2 opened the pull request:
This refers to https://github.com/hadley/httr/issues/363 where the reddit oauth2.0 example does not work unless you can add to the header.  In the case of reddit, you need to add the user agent, and the resulting example/demo works:

```r
app <- oauth_app("reddit", "bvmjj2EOBvOknQ", "n8ueSvTNdlE0BDDJpLljvmgUGUw")
reddit <- oauth_endpoint(
  authorize = "https://www.reddit.com/api/v1/authorize",
  access = "https://www.reddit.com/api/v1/access_token"
)
token <- oauth2.0_token(reddit, app, 
  scope = "read modposts", 
  use_basic_auth = TRUE,
  config = user_agent("redditr_agent"))
```

whereas the example does not (429 too many requests, because default user-agent is over-used):
```r
app <- oauth_app("reddit", "bvmjj2EOBvOknQ", "n8ueSvTNdlE0BDDJpLljvmgUGUw")
reddit <- oauth_endpoint(
  authorize = "https://www.reddit.com/api/v1/authorize",
  access = "https://www.reddit.com/api/v1/access_token"
)
token <- oauth2.0_token(reddit, app, 
  scope = "read modposts", 
  use_basic_auth = TRUE)
```

hadley commented on R/oauth-token.r:
```diff
@@ -197,10 +197,12 @@ oauth2.0_token <- function(endpoint, app, scope = NULL, user_params = NULL,
                            use_basic_auth = FALSE,
-                           cache = getOption("httr_oauth_cache")) {
+                           cache = getOption("httr_oauth_cache"),
+                           config = list()) {
```
I think this needs a more specific name because there are other places in the OAuth dance that could have config params. Maybe `config_init`?

hadley commented:
Fixes #413

muschellij2 commented:
Should be sufficient to merge after checks are done.  

* Fix reddit demo where `user_agent` was being overloaded.  Now can pass `user_agent` using `config_init` to `init_oauth2.0` and `Token2.0$new()`  (@muschellij2 @hadley #363 and #413).

* Changed example in `write_stream` from https://jeroenooms.github.io/data/diamonds.json to https://github.com/jeroen/data/raw/gh-pages/diamonds.json as this had moved (@muschellij2).

Also, fixed an example on `write_stream`: https://github.com/r-lib/httr/blob/1078b1d35d0c5d0a61d53461cc6c37f9fb49551d/R/write-function.R#L61 as that link was dead and causing errors in the "examples" part of R CMD check

hadley commented:
Did you close this deliberately?

muschellij2 commented:
Just meant to comment, not close and comment.

John

On Wed, Jul 26, 2017 at 9:54 AM, Hadley Wickham <notifications@github.com>
wrote:

> Did you close this deliberately?
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/r-lib/httr/pull/411#issuecomment-318059968>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABBnrk58V1xyY9iyVQdu6vXH4Xe0DGHNks5sR0UmgaJpZM4K4dz2>
> .
>

hadley commented:
Maybe it's better to resubmit anyway - your PR now has quite a few spurious changes and two concerns. That makes it hard to review efficiently.

muschellij2 commented:
Agreed - it was a few months ago so I think I may have some gaps there.
I'll check and resubmit

John

On Wed, Jul 26, 2017 at 10:59 AM, Hadley Wickham <notifications@github.com>
wrote:

> Maybe it's better to resubmit anyway - your PR now has quite a few
> spurious changes and two concerns. That makes it hard to review efficiently.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/r-lib/httr/pull/411#issuecomment-318080114>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABBnrkVB85EA9BBNW_619OeIbmYCQLB9ks5sR1RMgaJpZM4K4dz2>
> .
>
